File: app/src/main/java/com/raylabs/doggie/data/DoggieRepository.java
package com.raylabs.doggie.data;

import androidx.annotation.NonNull;
import androidx.lifecycle.LiveData;
import androidx.paging.Pager;
import androidx.paging.PagingConfig;
import androidx.paging.PagingData;

import com.raylabs.doggie.data.paging.BreedImagesPagingSource;
import com.raylabs.doggie.data.paging.CategoriesPagingSource;
import com.raylabs.doggie.data.source.local.LocalDataSource;
import com.raylabs.doggie.data.source.local.entity.DoggieEntity;
import com.raylabs.doggie.data.source.remote.ApiResponse;
import com.raylabs.doggie.data.source.remote.RemoteDataSource;
import com.raylabs.doggie.utils.AppExecutors;
import com.raylabs.doggie.vo.BreedCategory;
import com.raylabs.doggie.vo.Resource;

import java.util.ArrayList;
import java.util.List;

import kotlinx.coroutines.flow.Flow;

public class DoggieRepository implements DoggieDataSource {

    private volatile static DoggieRepository INSTANCE = null;
    private static final String TAG = "DoggieRepository";

    private static final int CATEGORY_PAGE_SIZE = 20;
    private static final int DEFAULT_BREED_IMAGE_PAGE_SIZE = 10;

    private final RemoteDataSource remoteDataSource;
    private final LocalDataSource localDataSource;
    private final AppExecutors appExecutors;

    private DoggieRepository(@NonNull RemoteDataSource remoteDataSource, LocalDataSource localDataSource, AppExecutors appExecutors) {
        this.remoteDataSource = remoteDataSource;
        this.localDataSource = localDataSource;
        this.appExecutors = appExecutors;
    }

    public static DoggieRepository getInstance(RemoteDataSource remoteDataSource, LocalDataSource localDataSource, AppExecutors appExecutors) {
        if (INSTANCE == null) {
            synchronized (DoggieRepository.class) {
                INSTANCE = new DoggieRepository(remoteDataSource, localDataSource, appExecutors);
            }
        }
        return INSTANCE;
    }

    @Override
    public LiveData<Resource<List<DoggieEntity>>> getAllImage(String countItem) {
        return new NetworkBoundResource<List<DoggieEntity>, List<String>>(appExecutors) {

            @Override
            protected LiveData<List<DoggieEntity>> loadFromDB() {
                return localDataSource.getAllDoggie("for-you");
            }

            @Override
            protected Boolean shouldFetch(List<DoggieEntity> data) {
                return (data == null) || (data.size() == 0);
            }

            @Override
            protected LiveData<ApiResponse<List<String>>> createCall() {
                return remoteDataSource.getAllImage(countItem);
            }

            /**
             * Regex For get Type Dog
             */
            @Override
            protected void saveCallResult(List<String> data) {
                ArrayList<DoggieEntity> doggieList = new ArrayList<>();
                for (String response : data) {
                    DoggieEntity doggie = new DoggieEntity(response.split("/")[4], response, "for-you");
                    doggieList.add(doggie);
                }
                localDataSource.insertDoggie(doggieList);
            }
        }.asLiveData();
    }

    @Override
    public LiveData<Resource<List<DoggieEntity>>> getLikedImage(String countItem) {
        return new NetworkBoundResource<List<DoggieEntity>, List<String>>(appExecutors) {

            @Override
            protected LiveData<List<DoggieEntity>> loadFromDB() {
                return localDataSource.getAllDoggie("liked");
            }

            @Override
            protected Boolean shouldFetch(List<DoggieEntity> data) {
                return (data == null) || (data.size() == 0);
            }

            @Override
            protected LiveData<ApiResponse<List<String>>> createCall() {
                return remoteDataSource.getAllImage(countItem);
            }

            /**
             * Regex For get Type Dog
             */
            @Override
            protected void saveCallResult(List<String> data) {
                ArrayList<DoggieEntity> doggieList = new ArrayList<>();
                for (String response : data) {
                    DoggieEntity doggie = new DoggieEntity(response.split("/")[4], response, "liked");
                    doggieList.add(doggie);
                }
                localDataSource.insertDoggie(doggieList);
            }
        }.asLiveData();
    }

    @Override
    public LiveData<Resource<List<DoggieEntity>>> getPopularImage(String countItem) {
        return new NetworkBoundResource<List<DoggieEntity>, List<String>>(appExecutors) {

            @Override
            protected LiveData<List<DoggieEntity>> loadFromDB() {
                return localDataSource.getAllDoggie("popular");
            }

            @Override
            protected Boolean shouldFetch(List<DoggieEntity> data) {
                return (data == null) || (data.size() == 0);
            }

            @Override
            protected LiveData<ApiResponse<List<String>>> createCall() {
                return remoteDataSource.getAllImage(countItem);
            }

            /**
             * Regex For get Type Dog
             */
            @Override
            protected void saveCallResult(List<String> data) {
                ArrayList<DoggieEntity> doggieList = new ArrayList<>();
                for (String response : data) {
                    DoggieEntity doggie = new DoggieEntity(response.split("/")[4], response, "popular");
                    doggieList.add(doggie);
                }
                localDataSource.insertDoggie(doggieList);
            }
        }.asLiveData();
    }

    @Override
    public Flow<PagingData<BreedCategory>> getCategories() {
        Pager<Integer, BreedCategory> pager = new Pager<>(
                new PagingConfig(
                        CATEGORY_PAGE_SIZE,
                        CATEGORY_PAGE_SIZE,
                        false
                ),
                () -> new CategoriesPagingSource(remoteDataSource, CATEGORY_PAGE_SIZE)
        );
        return pager.getFlow();
    }

    @Override
    public Flow<PagingData<String>> getBreedImages(String breed, String subBreed, int pageSize) {
        int effectivePageSize = pageSize > 0 ? pageSize : DEFAULT_BREED_IMAGE_PAGE_SIZE;
        Pager<Integer, String> pager = new Pager<>(
                new PagingConfig(
                        effectivePageSize,
                        effectivePageSize,
                        false
                ),
                () -> new BreedImagesPagingSource(remoteDataSource, breed, subBreed, effectivePageSize)
        );
        return pager.getFlow();
    }

}

File: app/src/main/java/com/raylabs/doggie/data/DoggieRepository.kt
package com.raylabs.doggie.data

import android.content.Context
import androidx.annotation.VisibleForTesting
import androidx.lifecycle.LiveData
import androidx.paging.Pager
import androidx.paging.PagingConfig
import androidx.paging.PagingData
import com.raylabs.doggie.data.paging.BreedImagesPagingSource
import com.raylabs.doggie.data.source.local.BreedCategoryLocalDataSource
import com.raylabs.doggie.data.source.local.LocalDataSource
import com.raylabs.doggie.data.source.local.entity.DoggieEntity
import com.raylabs.doggie.data.source.remote.ApiResponse
import com.raylabs.doggie.data.source.remote.RemoteDataSource
import com.raylabs.doggie.utils.AppExecutors
import com.raylabs.doggie.vo.BreedCategory
import com.raylabs.doggie.vo.Resource
import kotlinx.coroutines.flow.Flow

class DoggieRepository private constructor(
    context: Context,
    private val remoteDataSource: RemoteDataSource,
    private val localDataSource: LocalDataSource,
    breedCategoryLocalDataSource: BreedCategoryLocalDataSource,
    private val appExecutors: AppExecutors
) : DoggieDataSource {

    private val breedCategoryHelper = BreedCategoryRepositoryHelper(
        context,
        remoteDataSource,
        breedCategoryLocalDataSource,
        PagingConfig(CATEGORY_PAGE_SIZE, CATEGORY_PAGE_SIZE, false)
    )

    init {
        breedCategoryHelper.scheduleBackgroundSync()
    }

    override fun getAllImage(countItem: String): LiveData<Resource<List<DoggieEntity>>> {
        return object : NetworkBoundResource<List<DoggieEntity>, List<String>>(appExecutors) {
            override fun loadFromDB(): LiveData<List<DoggieEntity>> =
                localDataSource.getAllDoggie(FOR_YOU_TAG)

            override fun shouldFetch(data: List<DoggieEntity>?): Boolean = data.isNullOrEmpty()

            override fun createCall(): LiveData<ApiResponse<List<String>>> =
                remoteDataSource.getAllImage(countItem)

            override fun saveCallResult(data: List<String>) {
                val doggies = data.mapNotNull { link ->
                    val parts = link.split('/')
                    val type = parts.getOrNull(4) ?: return@mapNotNull null
                    DoggieEntity(type, link, FOR_YOU_TAG)
                }
                if (doggies.isNotEmpty()) {
                    localDataSource.insertDoggie(doggies)
                }
            }
        }.asLiveData()
    }

    override fun getLikedImage(countItem: String): LiveData<Resource<List<DoggieEntity>>> {
        return object : NetworkBoundResource<List<DoggieEntity>, List<String>>(appExecutors) {
            override fun loadFromDB(): LiveData<List<DoggieEntity>> =
                localDataSource.getAllDoggie(LIKED_TAG)

            override fun shouldFetch(data: List<DoggieEntity>?): Boolean = data.isNullOrEmpty()

            override fun createCall(): LiveData<ApiResponse<List<String>>> =
                remoteDataSource.getAllImage(countItem)

            override fun saveCallResult(data: List<String>) {
                val doggies = data.mapNotNull { link ->
                    val parts = link.split('/')
                    val type = parts.getOrNull(4) ?: return@mapNotNull null
                    DoggieEntity(type, link, LIKED_TAG)
                }
                if (doggies.isNotEmpty()) {
                    localDataSource.insertDoggie(doggies)
                }
            }
        }.asLiveData()
    }

    override fun getPopularImage(countItem: String): LiveData<Resource<List<DoggieEntity>>> {
        return object : NetworkBoundResource<List<DoggieEntity>, List<String>>(appExecutors) {
            override fun loadFromDB(): LiveData<List<DoggieEntity>> =
                localDataSource.getAllDoggie(POPULAR_TAG)

            override fun shouldFetch(data: List<DoggieEntity>?): Boolean = data.isNullOrEmpty()

            override fun createCall(): LiveData<ApiResponse<List<String>>> =
                remoteDataSource.getAllImage(countItem)

            override fun saveCallResult(data: List<String>) {
                val doggies = data.mapNotNull { link ->
                    val parts = link.split('/')
                    val type = parts.getOrNull(4) ?: return@mapNotNull null
                    DoggieEntity(type, link, POPULAR_TAG)
                }
                if (doggies.isNotEmpty()) {
                    localDataSource.insertDoggie(doggies)
                }
            }
        }.asLiveData()
    }

    override fun getCategories(): Flow<PagingData<BreedCategory>> = breedCategoryHelper.pager()

    override fun requestCategoryPreview(category: BreedCategory) {
        breedCategoryHelper.requestPreview(category)
    }

    override fun getBreedImages(
        breed: String,
        subBreed: String?,
        pageSize: Int
    ): Flow<PagingData<String>> {
        val effectivePageSize = pageSize.takeIf { it > 0 } ?: DEFAULT_BREED_IMAGE_PAGE_SIZE
        return Pager(
            PagingConfig(
                pageSize = effectivePageSize,
                initialLoadSize = effectivePageSize,
                enablePlaceholders = false
            )
        ) {
            BreedImagesPagingSource(remoteDataSource, breed, subBreed, effectivePageSize)
        }.flow
    }

    override fun observeBreedCatalog(): Flow<List<BreedCategory>> =
        breedCategoryHelper.catalogFlow()

    companion object {
        private const val CATEGORY_PAGE_SIZE = 20
        private const val DEFAULT_BREED_IMAGE_PAGE_SIZE = 10
        private const val FOR_YOU_TAG = "for-you"
        private const val LIKED_TAG = "liked"
        private const val POPULAR_TAG = "popular"

        @Volatile
        private var INSTANCE: DoggieRepository? = null

        @JvmStatic
        fun getInstance(
            context: Context,
            remoteDataSource: RemoteDataSource,
            localDataSource: LocalDataSource,
            breedCategoryLocalDataSource: BreedCategoryLocalDataSource,
            appExecutors: AppExecutors
        ): DoggieRepository {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: DoggieRepository(
                    context.applicationContext,
                    remoteDataSource,
                    localDataSource,
                    breedCategoryLocalDataSource,
                    appExecutors
                ).also { INSTANCE = it }
            }
        }

        @VisibleForTesting
        @JvmStatic
        fun resetForTesting() {
            INSTANCE = null
        }
    }
}
